File: test/performance/data-structures/hash/hash-map.test.ts
import { HashMap } from '../../../../src';
import { HashMap as CHashMap } from 'js-sdsl';
import * as Benchmark from 'benchmark';
import { magnitude } from '../../../utils';
import { isCompetitor } from '../../../config';

const suite = new Benchmark.Suite();
const { MILLION } = magnitude;

suite.add(`${MILLION.toLocaleString()} set`, () => {
  const hm = new HashMap<number, number>();

  for (let i = 0; i < MILLION; i++) hm.set(i, i);
});

if (isCompetitor) {
  suite.add(`CPT ${MILLION.toLocaleString()} set`, () => {
    const hm = new CHashMap<number, number>();

    for (let i = 0; i < MILLION; i++) hm.setElement(i, i);
  });
}

suite.add(`Native JS Map ${MILLION.toLocaleString()} set`, () => {
  const hm = new Map<number, number>();

  for (let i = 0; i < MILLION; i++) hm.set(i, i);
});

suite.add(`Native JS Set ${MILLION.toLocaleString()} add`, () => {
  const hs = new Set<number>();

  for (let i = 0; i < MILLION; i++) hs.add(i);
});

suite.add(`${MILLION.toLocaleString()} set & get`, () => {
  const hm = new HashMap<number, number>();

  for (let i = 0; i < MILLION; i++) hm.set(i, i);
  for (let i = 0; i < MILLION; i++) hm.get(i);
});

if (isCompetitor) {
  suite.add(`CPT ${MILLION.toLocaleString()} set & get`, () => {
    const hm = new CHashMap<number, number>();

    for (let i = 0; i < MILLION; i++) hm.setElement(i, i);
    for (let i = 0; i < MILLION; i++) hm.getElementByKey(i);
  });
}

suite.add(`Native JS Map ${MILLION.toLocaleString()} set & get`, () => {
  const hm = new Map<number, number>();

  for (let i = 0; i < MILLION; i++) hm.set(i, i);
  for (let i = 0; i < MILLION; i++) hm.get(i);
});

suite.add(`Native JS Set ${MILLION.toLocaleString()} add & has`, () => {
  const hs = new Set<number>();

  for (let i = 0; i < MILLION; i++) hs.add(i);
  for (let i = 0; i < MILLION; i++) hs.has(i);
});

suite.add(`${MILLION.toLocaleString()} ObjKey set & get`, () => {
  const hm = new HashMap<[number, number], number>();
  const objKeys: [number, number][] = [];
  for (let i = 0; i < MILLION; i++) {
    const obj: [number, number] = [i, i];
    objKeys.push(obj);
    hm.set(obj, i);
  }
  for (let i = 0; i < MILLION; i++) hm.get(objKeys[i]);
});

suite.add(`Native JS Map ${MILLION.toLocaleString()} ObjKey set & get`, () => {
  const hm = new Map<[number, number], number>();
  const objs: [number, number][] = [];
  for (let i = 0; i < MILLION; i++) {
    const obj: [number, number] = [i, i];
    objs.push(obj);
    hm.set(obj, i);
  }
  for (let i = 0; i < MILLION; i++) hm.get(objs[i]);
});

suite.add(`Native JS Set ${MILLION.toLocaleString()} ObjKey add & has`, () => {
  const hs = new Set<[number, number]>();
  const objs: [number, number][] = [];
  for (let i = 0; i < MILLION; i++) {
    const obj: [number, number] = [i, i];
    objs.push(obj);
    hs.add(obj);
  }
  for (let i = 0; i < MILLION; i++) hs.has(objs[i]);
});

export { suite };

File: test/performance/data-structures/hash/hash-map.test.cpp
#include <benchmark/benchmark.h>
#include <unordered_map>
#include <unordered_set>

const int MILLION = 1000000;

std::string formatNumber(int num) {
  std::string result = std::to_string(num);
  int n = result.length() - 3;
  while (n > 0) {
    result.insert(n, ",");
    n -= 3;
  }
  return result;
}

// Test 1: 1M set
static void Set1M(benchmark::State& state) {
  for (auto _ : state) {
    std::unordered_map<int, int> hm;
    for (int i = 0; i < MILLION; i++) {
      hm[i] = i;
    }
    benchmark::DoNotOptimize(hm);
  }
}
BENCHMARK(Set1M)->Name("1M set");

// Test 2: Native std::map 1M set (for comparison)
static void StdMapSet1M(benchmark::State& state) {
  for (auto _ : state) {
    std::map<int, int> hm;
    for (int i = 0; i < MILLION; i++) {
      hm[i] = i;
    }
    benchmark::DoNotOptimize(hm);
  }
}
BENCHMARK(StdMapSet1M)->Name("std::map 1M set");

// Test 3: std::unordered_set 1M add
static void UnorderedSetAdd1M(benchmark::State& state) {
  for (auto _ : state) {
    std::unordered_set<int> hs;
    for (int i = 0; i < MILLION; i++) {
      hs.insert(i);
    }
    benchmark::DoNotOptimize(hs);
  }
}
BENCHMARK(UnorderedSetAdd1M)->Name("std::unordered_set 1M add");

// Test 4: 1M set & get
static void SetAndGet1M(benchmark::State& state) {
  for (auto _ : state) {
    std::unordered_map<int, int> hm;
    for (int i = 0; i < MILLION; i++) {
      hm[i] = i;
    }
    int count = 0;
    for (int i = 0; i < MILLION; i++) {
      if (hm.find(i) != hm.end()) {
        count++;
      }
    }
    benchmark::DoNotOptimize(count);
  }
}
BENCHMARK(SetAndGet1M)->Name("1M set & get");

// Test 5: std::map 1M set & get
static void StdMapSetAndGet1M(benchmark::State& state) {
  for (auto _ : state) {
    std::map<int, int> hm;
    for (int i = 0; i < MILLION; i++) {
      hm[i] = i;
    }
    int count = 0;
    for (int i = 0; i < MILLION; i++) {
      if (hm.find(i) != hm.end()) {
        count++;
      }
    }
    benchmark::DoNotOptimize(count);
  }
}
BENCHMARK(StdMapSetAndGet1M)->Name("std::map 1M set & get");

// Test 6: std::unordered_set 1M add & has
static void UnorderedSetAddAndHas1M(benchmark::State& state) {
  for (auto _ : state) {
    std::unordered_set<int> hs;
    for (int i = 0; i < MILLION; i++) {
      hs.insert(i);
    }
    int count = 0;
    for (int i = 0; i < MILLION; i++) {
      if (hs.find(i) != hs.end()) {
        count++;
      }
    }
    benchmark::DoNotOptimize(count);
  }
}
BENCHMARK(UnorderedSetAddAndHas1M)->Name("std::unordered_set 1M add & has");

// Test 7: 1M ObjKey set & get
static void ObjKeySetAndGet1M(benchmark::State& state) {
  for (auto _ : state) {
    struct Obj { int key; int value; };
    struct ObjHash { size_t operator()(const Obj& o) const { return std::hash<int>()(o.key); } };
    struct ObjEqual { bool operator()(const Obj& a, const Obj& b) const { return a.key == b.key; } };
    
    std::unordered_map<Obj, int, ObjHash, ObjEqual> hm;
    std::vector<Obj> objKeys;
    for (int i = 0; i < MILLION; i++) {
      Obj obj = {i, i};
      objKeys.push_back(obj);
      hm[obj] = i;
    }
    
    int count = 0;
    for (int i = 0; i < MILLION; i++) {
      if (hm.find(objKeys[i]) != hm.end()) {
        count++;
      }
    }
    benchmark::DoNotOptimize(count);
  }
}
BENCHMARK(ObjKeySetAndGet1M)->Name("1M ObjKey set & get");

// Test 8: std::map 1M ObjKey set & get
static void StdMapObjKeySetAndGet1M(benchmark::State& state) {
  for (auto _ : state) {
    struct Obj { int key; int value; bool operator<(const Obj& other) const { return key < other.key; } };
    
    std::map<Obj, int> hm;
    std::vector<Obj> objs;
    for (int i = 0; i < MILLION; i++) {
      Obj obj = {i, i};
      objs.push_back(obj);
      hm[obj] = i;
    }
    
    int count = 0;
    for (int i = 0; i < MILLION; i++) {
      if (hm.find(objs[i]) != hm.end()) {
        count++;
      }
    }
    benchmark::DoNotOptimize(count);
  }
}
BENCHMARK(StdMapObjKeySetAndGet1M)->Name("std::map 1M ObjKey set & get");

// Test 9: std::unordered_set 1M ObjKey add & has
static void UnorderedSetObjKeyAddAndHas1M(benchmark::State& state) {
  for (auto _ : state) {
    struct Obj { int key; int value; };
    struct ObjHash { size_t operator()(const Obj& o) const { return std::hash<int>()(o.key); } };
    struct ObjEqual { bool operator()(const Obj& a, const Obj& b) const { return a.key == b.key; } };
    
    std::unordered_set<Obj, ObjHash, ObjEqual> hs;
    std::vector<Obj> objs;
    for (int i = 0; i < MILLION; i++) {
      Obj obj = {i, i};
      objs.push_back(obj);
      hs.insert(obj);
    }
    
    int count = 0;
    for (int i = 0; i < MILLION; i++) {
      if (hs.find(objs[i]) != hs.end()) {
        count++;
      }
    }
    benchmark::DoNotOptimize(count);
  }
}
BENCHMARK(UnorderedSetObjKeyAddAndHas1M)->Name("std::unordered_set 1M ObjKey add & has");

BENCHMARK_MAIN();
